File: pkg/agent/core/ngt/handler/grpc/flush.go
// Copyright (C) 2019-2024 vdaas.org vald team <vald@vdaas.org>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//	https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package grpc

import (
	"context"
	"fmt"
	"sync/atomic"

	"github.com/vdaas/vald/apis/grpc/v1/payload"
	"github.com/vdaas/vald/apis/grpc/v1/vald"
	"github.com/vdaas/vald/internal/errors"
	"github.com/vdaas/vald/internal/info"
	"github.com/vdaas/vald/internal/log"
	"github.com/vdaas/vald/internal/net/grpc/errdetails"
	"github.com/vdaas/vald/internal/net/grpc/status"
	"github.com/vdaas/vald/internal/observability/trace"
	"go.opentelemetry.io/otel/attribute"
)

// Flush removes all vectors that are indexed and uncommitted in the `vald-agent`.
func (s *server) Flush(ctx context.Context, req *payload.Flush_Request) (*payload.Info_Index_Count, error) {
	_, span := trace.StartSpan(ctx, apiName+"/"+vald.FlushRPCName)
	defer func() {
		if span != nil {
			span.End()
		}
	}()
	err := s.ngt.RegenerateIndexes(ctx)
	if err != nil {
		var attrs []attribute.KeyValue
		if errors.Is(err, errors.ErrFlushingIsInProgress) {
			err = status.WrapWithAborted("Flush API aborted to process search request due to flushing indices is in progress", err,
				&errdetails.RequestInfo{
					ServingData: errdetails.Serialize(req),
				},
				&errdetails.ResourceInfo{
					ResourceType: ngtResourceType + "/ngt.Flush",
					ResourceName: fmt.Sprintf("%s: %s(%s)", apiName, s.name, s.ip),
				})
			log.Debug(err)
			attrs = trace.StatusCodeAborted(err.Error())
		} else {
			err = status.WrapWithInternal("Flush API failed", err,
				&errdetails.RequestInfo{
					ServingData: errdetails.Serialize(req),
				},
				&errdetails.ResourceInfo{
					ResourceType: ngtResourceType + "/ngt.Flush",
					ResourceName: fmt.Sprintf("%s: %s(%s)", apiName, s.name, s.ip),
				}, info.Get())
			log.Error(err)
			attrs = trace.StatusCodeInternal(err.Error())
		}
		if span != nil {
			span.RecordError(err)
			span.SetAttributes(attrs...)
			span.SetStatus(trace.StatusError, err.Error())
		}
		return nil, err
	}

	var (
		stored     uint32
		uncommited uint32
		indexing   atomic.Value
		saving     atomic.Value
	)
	stored = 0
	uncommited = 0
	indexing.Store(false)
	saving.Store(false)

	cnts := &payload.Info_Index_Count{
		Stored:      atomic.LoadUint32(&stored),
		Uncommitted: atomic.LoadUint32(&uncommited),
		Indexing:    indexing.Load().(bool),
		Saving:      saving.Load().(bool),
	}

	return cnts, nil
}

File: rust/bin/agent/src/handler/flush.rs
//
// Copyright (C) 2019-2026 vdaas.org vald team <vald@vdaas.org>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

use algorithm::Error;
use log::{debug, error, info};
use prost::Message;
use proto::{payload::v1::info, vald::v1::flush_server};
use tonic::{Code, Status};
use tonic_types::StatusExt;

use crate::handler::common::build_error_details;

#[tonic::async_trait]
impl<S: algorithm::ANN + 'static> flush_server::Flush for super::Agent<S> {
    async fn flush(
        &self,
        request: tonic::Request<proto::payload::v1::flush::Request>,
    ) -> std::result::Result<tonic::Response<proto::payload::v1::info::index::Count>, Status> {
        info!("Recieved a request from {:?}", request.remote_addr());
        let hostname = cargo::util::hostname()?;
        let domain = hostname.to_str().unwrap();
        {
            let mut s = self.s.write().await;
            let result = s.regenerate_indexes().await;
            match result {
                Err(err) => {
                    let resource_type = self.resource_type.clone() + "/qbg.Flush";
                    let resource_name = format!("{}: {}({})", self.api_name, self.name, self.ip);
                    let err_details = build_error_details(
                        err.to_string(),
                        domain,
                        "",
                        request.get_ref().encode_to_vec(),
                        &resource_type,
                        &resource_name,
                        None,
                    );
                    let status = match err {
                        Error::FlushingIsInProgress {} => {
                            let status = Status::with_error_details(Code::Aborted, "Flush API aborted due to flushing indices is in progress", err_details);
                            debug!("{:?}", status);
                            status
                        }
                        Error::WriteOperationToReadReplica {} => {
                            let status = Status::with_error_details(
                                Code::Aborted,
                                "Flush API aborted due to agent is read only",
                                err_details,
                            );
                            debug!("{:?}", status);
                            status
                        }
                        _ => {
                            let status = Status::with_error_details(
                                Code::Internal,
                                "Flush API is failed",
                                err_details,
                            );
                            error!("{:?}", status);
                            status
                        }
                    };
                    Err(status)
                }
                Ok(()) => {
                    let res = info::index::Count {
                        stored: 0,
                        uncommitted: 0,
                        indexing: false,
                        saving: false,
                    };
                    Ok(tonic::Response::new(res))
                }
            }
        }
    }
}
